### ISSUE TITLE
wchild_, getDynamicChild, etc should allow returning a Deferred

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11476]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)| @jyknight reported|
|-|-|
|Trac ID|trac#344|
|Type|enhancement|
|Created|2003-10-16 05:13:57Z|

<details><summary>Searchable metadata</summary>

```
trac-id__344 344
type__enhancement enhancement
reporter__jknight jknight
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web web
keywords__ 
time__1066281237000000 1066281237000000
changetime__1468795778304213 1468795778304213
version__ 
owner__ 
cc__radix cc__itamarst cc__dp cc__moshez cc__jknight
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
At least woven's wchild_ and getDynamicChild should allow returning a Deferred, which will get 
autowrapped in a DeferredResource.

Additionally, as many of the other get*Child* methods as make sense/are possible should similarly
work. (aka: if it's possible to put the autowrapping outside of woven-specific code that'd be extra-
nifty-bonus-points)

[18:56] &lt;stranger> (it would be nice if one could return a deferred that would become a page 
object...)
[18:57] &lt;foom> returning a deferred from a wchild_ doesn't work? if not then return a 
DeferredResource
[18:57] &lt;itamar> foom: don't think it works, no
[18:57] &lt;itamar> though it probably ought to
[18:57] &lt;itamar> add an issue
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="moshez's avatar" src="https://avatars.githubusercontent.com/u/422703?s=50" width="50" height="50">](https://github.com/moshez)|@moshez commented|
|-|-|

```
#!html
<pre>
They will when getChildWithDefault will allow returning a deferred,
and there's already a feature filed for that.
Why on earth should there be separate code in Woven?
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
I think maybe I'm blind, because I don't see the other feature request filed? 

Also, my request was for a behavior, not an implementation. As I tried to say, if it makes the most 
sense to put the code outside woven, great!
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
1. The feature request is for render() AFAICT.

2. wchild_ may be be done in render(), for example, who knows, in which case
it'd need woven specific code. Probably not, but this is Woven, where EVERYTHING
IS POSSIBLE.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)|@itamarst commented|
|-|-|

```
#!html
<pre>
Now that we have render_GET etc., they should be altered so that returning
Deferred is supported. Someone should investigate whether adding support for
getChild will break backwards compatability - my guess is that with the
appropriate checks in the resource-traversal loop it should be fine. Donovan
should probably be able to help with is.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="radix's avatar" src="https://avatars.githubusercontent.com/u/227068?s=50" width="50" height="50">](https://github.com/radix)|@radix commented|
|-|-|

```
#!html
<pre>
WONTFIX; this stuff is already working in nevow.
</pre>
```

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="GitHub's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|GitHub commented|
|-|-|

In changeset 6e0937997ced5a2a7860afd74fb38e6ca2085205
```
#!CommitTicketReference repository="" revision="6e0937997ced5a2a7860afd74fb38e6ca2085205"
Merge pull request #344 from twisted/8468-rodrigc-exc_clear

Author: rodrigc
Reviewer: markrwilliams
Fixes: #8468

Do not call sys.exc_clear() on Python 3 in twisted/scripts/trial.py
```
